File: rekall-domain/src/main/java/dev/rekall/domain/Task.java
package dev.rekall.domain;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.OneToOne;
import jakarta.persistence.OrderBy;
import jakarta.persistence.OrderColumn;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Pattern;
import jakarta.validation.constraints.Size;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

@Entity
@Table(
        name = "task",
        uniqueConstraints = @UniqueConstraint(name = "uq_task_project_label", columnNames = {"project_id", "label"}))
@Getter
public class Task {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "id", nullable = false, updatable = false)
    private UUID id;

    @NotBlank
    @Size(max = 160)
    @Pattern(regexp = Slug.PATTERN, message = "must be a slug: lowercase letters, digits, '-', '_' or '.'")
    @Column(name = "label", nullable = false, length = 160)
    @Setter
    private String label;

    @NotBlank
    @Size(max = 200)
    @Column(name = "title", nullable = false, length = 200)
    @Setter
    private String title;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 20)
    @Setter
    private TaskStatus status = TaskStatus.TODO;

    @Column(name = "description", length = 100_000)
    @Setter
    private String description;

    @Column(name = "auto_wrapup", nullable = false)
    private boolean autoWrapup = false;

    @Size(max = 2_000)
    @Column(name = "wrapup_directive", length = 2_000)
    private String wrapupDirective;

    @Enumerated(EnumType.STRING)
    @Column(name = "review_state", nullable = false, length = 20)
    private TaskStepState reviewState = TaskStepState.OPEN;

    @Column(name = "claimed_at")
    private Instant claimedAt;

    @Column(name = "accepted_at")
    private Instant acceptedAt;

    @Size(max = 2_000)
    @Column(name = "review_note", length = 2_000)
    @Setter
    private String reviewNote;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "project_id", nullable = false, foreignKey = @jakarta.persistence.ForeignKey(name = "fk_task_project"))
    @Setter
    private Project project;

    @ManyToMany
    @JoinTable(
            name = "document_task",
            joinColumns = @JoinColumn(name = "task_id", foreignKey = @jakarta.persistence.ForeignKey(name = "fk_document_task_task")),
            inverseJoinColumns = @JoinColumn(name = "document_id", foreignKey = @jakarta.persistence.ForeignKey(name = "fk_document_task_document")))
    @OrderColumn(name = "position")
    private List<Document> documents = new ArrayList<>();

    @OneToMany(mappedBy = "task", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("position ASC")
    private List<TaskStep> steps = new ArrayList<>();

    @OneToOne(mappedBy = "task", cascade = CascadeType.ALL, orphanRemoval = true)
    @Setter
    private Wrapup wrapup;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;

    protected Task() {
    }

    public Task(String label, String title) {
        this.label = label;
        this.title = title;
    }

    /**
     * True while this task has no checklist. The task-scoped review line
     * ({@link #reviewState} and its two moments) only means anything here: once
     * one step is past {@code DRAFT} the checklist is the source of truth and
     * these columns are kept but ignored. Draft steps are a planning surface, not
     * a checklist, so a task that only holds drafts is still review-active.
     */
    public boolean reviewActive() {
        return steps.stream().allMatch(step -> step.getState() == TaskStepState.DRAFT);
    }

    /**
     * Walk the task-scoped review line, the same four values a {@link TaskStep}
     * moves along read at task scope: {@code OPEN} before a session has run or
     * after a send back, {@code RUNNING} while a session is attached to the task
     * anchor with no step target, {@code CLAIMED} once a Claude-authored wrapup
     * lands, {@code DONE} when the console accepts it. Each move rewrites the two
     * moments and the send-back note so they never outlive the state that set
     * them.
     */
    public void markReviewState(TaskStepState next) {
        if (reviewState == next) {
            return;
        }
        reviewState = next;
        Instant now = Instant.now();
        switch (next) {
            case OPEN, RUNNING -> {
                claimedAt = null;
                acceptedAt = null;
                reviewNote = null;
            }
            case CLAIMED -> {
                claimedAt = now;
                acceptedAt = null;
                reviewNote = null;
            }
            case DONE -> {
                if (claimedAt == null) {
                    claimedAt = now;
                }
                acceptedAt = now;
                reviewNote = null;
            }
        }
    }

    /**
     * The standing wrapup instruction for this task. While {@code auto} is on,
     * every {@code /rk ... wrapup} on the task or one of its steps is expected to
     * run without the console asking for it, folding in {@code directive} as the
     * wording it should follow. A blank directive is stored as none, and the
     * directive is cleared whenever the toggle goes off so a stale instruction
     * never outlives the intent that set it.
     */
    public void configureWrapup(boolean auto, String directive) {
        this.autoWrapup = auto;
        String trimmed = directive == null || directive.isBlank() ? null : directive.trim();
        this.wrapupDirective = auto ? trimmed : null;
    }

    public void attach(Document document) {
        if (documents.contains(document)) {
            return;
        }
        documents.add(document);
        document.getTasks().add(this);
    }

    public void detach(Document document) {
        if (documents.remove(document)) {
            document.getTasks().remove(this);
        }
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        return other instanceof Task that && id != null && id.equals(that.id);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }

    @Override
    public String toString() {
        return "Task[" + label + "]";
    }
}

File: rekall-model/src/task.rs
//! A unit of work inside a project. `label` is unique per project and is the anchor; `title` is
//! free text. A task with no checklist (or only draft steps) carries its own review line in
//! `review_state`, `claimed_at`, `accepted_at` and `review_note`.

use rekall_common::{Id, Instant, RekallError};
use sea_orm::entity::prelude::*;

use crate::constraints::{Phase, Violations};
use crate::enums::{TaskStatus, TaskStepState};

#[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
#[sea_orm(table_name = "task")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    pub id: Id,
    pub label: String,
    pub title: String,
    pub status: TaskStatus,
    pub description: Option<String>,
    pub review_state: TaskStepState,
    pub claimed_at: Option<Instant>,
    pub accepted_at: Option<Instant>,
    pub review_note: Option<String>,
    pub project_id: Id,
    pub tag_id: Option<Id>,
    pub created_at: Instant,
    pub updated_at: Instant,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(
        belongs_to = "super::project::Entity",
        from = "Column::ProjectId",
        to = "super::project::Column::Id",
        on_delete = "Cascade"
    )]
    Project,
    #[sea_orm(
        belongs_to = "super::tag::Entity",
        from = "Column::TagId",
        to = "super::tag::Column::Id",
        on_delete = "SetNull"
    )]
    Tag,
    #[sea_orm(has_many = "super::task_step::Entity")]
    TaskStep,
    #[sea_orm(has_one = "super::wrapup::Entity")]
    Wrapup,
    #[sea_orm(has_many = "super::document_task::Entity")]
    DocumentTask,
}

impl Related<super::project::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Project.def()
    }
}

impl Related<super::tag::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Tag.def()
    }
}

impl Related<super::task_step::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::TaskStep.def()
    }
}

impl Related<super::wrapup::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Wrapup.def()
    }
}

impl Related<super::document::Entity> for Entity {
    fn to() -> RelationDef {
        super::document_task::Relation::Document.def()
    }

    fn via() -> Option<RelationDef> {
        Some(super::document_task::Relation::Task.def().rev())
    }
}

impl ActiveModelBehavior for ActiveModel {}

/// `Task.reviewActive()`: true while the task has no checklist, only draft steps counting as
/// none. Once a step is past `DRAFT` the checklist is the source of truth and the review columns
/// are ignored.
pub fn review_active<'a>(step_states: impl IntoIterator<Item = &'a TaskStepState>) -> bool {
    step_states.into_iter().all(|state| *state == TaskStepState::Draft)
}

impl Model {
    /// A new task as `new Task(label, title)` left it: `TODO`, review line `OPEN`, no moments.
    pub fn new(label: String, title: String, project_id: Id, now: Instant) -> Self {
        Self {
            id: Id::random(),
            label,
            title,
            status: TaskStatus::Todo,
            description: None,
            review_state: TaskStepState::Open,
            claimed_at: None,
            accepted_at: None,
            review_note: None,
            project_id,
            tag_id: None,
            created_at: now,
            updated_at: now,
        }
    }

    /// `Task.markReviewState`: move the task-scoped review line. `OPEN` before a run or after a
    /// send back, `RUNNING` during a session, `CLAIMED` once a wrapup lands, `DONE` on accept.
    /// Each move resets the claim and accept moments and the note to match; marking the state
    /// the task is already in changes nothing.
    pub fn mark_review_state(&mut self, next: TaskStepState) {
        if self.review_state == next {
            return;
        }
        self.review_state = next;
        let now = Instant::now();
        match next {
            TaskStepState::Open | TaskStepState::Running => {
                self.claimed_at = None;
                self.accepted_at = None;
                self.review_note = None;
            }
            TaskStepState::Claimed => {
                self.claimed_at = Some(now);
                self.accepted_at = None;
                self.review_note = None;
            }
            TaskStepState::Done => {
                if self.claimed_at.is_none() {
                    self.claimed_at = Some(now);
                }
                self.accepted_at = Some(now);
                self.review_note = None;
            }
            // DRAFT is a step's state, never the task's; the switch in Java had no case for it.
            TaskStepState::Draft => {}
        }
    }

    pub fn anchor(&self, project_label: &str) -> String {
        format!("project:{project_label} task:{}", self.label)
    }

    pub fn validate(&self, phase: Phase) -> Result<(), RekallError> {
        Violations::new("Task", phase)
            .not_blank("label", Some(&self.label))
            .size("label", Some(&self.label), 160)
            .slug("label", Some(&self.label))
            .not_blank("title", Some(&self.title))
            .size("title", Some(&self.title), 200)
            .size("reviewNote", self.review_note.as_deref(), 2_000)
            .column("description", self.description.as_deref(), 100_000)
            .finish()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn new_task() -> Model {
        Model::new("in-app-claude".into(), "In-app Claude execution".into(), Id::random(), Instant::now())
    }

    #[test]
    fn a_fresh_task_opens_open_review_active_with_nothing_behind_it() {
        let task = new_task();
        assert_eq!(task.review_state, TaskStepState::Open);
        assert!(review_active(&[]));
        assert!(task.claimed_at.is_none());
        assert!(task.accepted_at.is_none());
        assert!(task.review_note.is_none());
    }

    #[test]
    fn running_then_claimed_then_accepted_each_move_stamps_its_own_moment() {
        let mut task = new_task();
        task.mark_review_state(TaskStepState::Running);
        assert!(task.claimed_at.is_none());
        assert!(task.accepted_at.is_none());

        task.mark_review_state(TaskStepState::Claimed);
        assert!(task.claimed_at.is_some());
        assert!(task.accepted_at.is_none());

        task.mark_review_state(TaskStepState::Done);
        assert!(task.accepted_at.is_some());
        assert!(task.claimed_at.is_some(), "the claim moment is kept when the work is later accepted");
    }

    #[test]
    fn accepting_straight_from_open_still_records_that_it_must_have_been_claimed() {
        let mut task = new_task();
        task.mark_review_state(TaskStepState::Done);
        assert!(task.claimed_at.is_some());
        assert!(task.accepted_at.is_some());
    }

    #[test]
    fn a_send_back_note_is_kept_while_open_and_dropped_the_moment_it_is_claimed_again() {
        let mut task = new_task();
        task.mark_review_state(TaskStepState::Open);
        task.review_note = Some("the export column is still wrong".into());
        assert_eq!(task.review_note.as_deref(), Some("the export column is still wrong"));

        task.mark_review_state(TaskStepState::Claimed);
        assert!(task.review_note.is_none());
    }

    #[test]
    fn sending_back_clears_every_moment() {
        let mut task = new_task();
        task.mark_review_state(TaskStepState::Running);
        task.mark_review_state(TaskStepState::Claimed);
        task.mark_review_state(TaskStepState::Done);

        task.mark_review_state(TaskStepState::Open);

        assert_eq!(task.review_state, TaskStepState::Open);
        assert!(task.claimed_at.is_none());
        assert!(task.accepted_at.is_none());
    }

    #[test]
    fn marking_the_state_it_is_already_in_changes_nothing() {
        let mut task = new_task();
        task.mark_review_state(TaskStepState::Claimed);
        let claimed_at = task.claimed_at;
        std::thread::sleep(std::time::Duration::from_millis(2));
        task.mark_review_state(TaskStepState::Claimed);
        assert_eq!(task.claimed_at, claimed_at);
    }

    #[test]
    fn only_drafts_count_as_no_checklist() {
        assert!(review_active(&[TaskStepState::Draft, TaskStepState::Draft]));
        assert!(!review_active(&[TaskStepState::Draft, TaskStepState::Open]));
    }
}
